lib/spack/spack/cmd/checkbinaries.py

```diff
@@ -0,0 +1,247 @@
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from __future__ import print_function
+
+from six.moves.urllib.request import urlopen
+from six.moves.urllib.error import URLError
+
+import json
+import os
+import re
+import yaml
+
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.config
+from spack.error import SpackError
+from spack.spec import Spec
+from spack.util.executable import which
+
+
+description = "check if binaries need to be rebuilt"
+section = "developer"
+level = "long"
+
+
+_buildcache_path = os.path.join('build_cache')
+_pkg_match_regex = re.compile(r'packages[/\\]([^/\\]+)[/\\]package\.py')
+_pkg_yaml_match_regex = re.compile(r'<a href=[^>]+\>\s*([^<]+)\s*\<')
+
+
+def changed_package_files(args):
+    """Get list of changed package.py files in the Spack repository."""
+
+    git = which('git', required=True)
+
+    range = "{0}...".format(args.base)
+
+    git_args = [
+        # Add changed files committed since branching off of develop
+        ['diff', '--name-only', '--diff-filter=ACMR', range],
+        # Add changed files that have been staged but not yet committed
+        ['diff', '--name-only', '--diff-filter=ACMR', '--cached'],
+        # Add changed files that are unstaged
+        ['diff', '--name-only', '--diff-filter=ACMR'],
+    ]
+
+    changed = set()
+
+    for arg_list in git_args:
+        files = git(*arg_list, output=str).split('\n')
+
+        for f in files:
+            # Look only for modified package.py files
+            m = _pkg_match_regex.search(f)
+            if m:
+                changed.add(m.group(1))
+
+    return sorted(changed)
+
+
+def compute_local_hash(pkg_name, spec_yaml):
+    local_spec = Spec.from_yaml(spec_yaml)
+    tty.msg('local dag hash for %s: %s' % (pkg_name, local_spec.dag_hash()))
+    local_spec.concretize()
+    return local_spec.full_hash()
+
+
+def compute_remote_hash(pkg_name, spec_yaml):
+    parsed_yaml = yaml.load(spec_yaml)
+    if 'full_hash' in parsed_yaml:
+        return parsed_yaml['full_hash']
+
+    msg = 'Binary mirror yaml file is missing full_hash attribute, aborting'
+    raise SpackError(msg)
+
+
+def check_needs_rebuild(pkg_name, spec_yaml):
+    local_pkg_hash = compute_local_hash(pkg_name, spec_yaml)
+    remote_pkg_hash = compute_remote_hash(pkg_name, spec_yaml)
+
+    tty.msg('  %s' % pkg_name)
+    tty.msg('    local hash: %s' % local_pkg_hash)
+    tty.msg('    remote hash: %s' % remote_pkg_hash)
+
+    if local_pkg_hash != remote_pkg_hash:
+        return True
+
+    return False
+
+
+def parse_yaml_list(index_html_contents):
+    result = []
+    match_list = _pkg_yaml_match_regex.findall(index_html_contents)
```

Comment:
This looks good but I think adding a parseable json file alongside index.html would be good.

Reply:
Yes, I agree.  While I did add a yaml index to the `buildcache` command, I didn't yet update this to read it.  Based on your comments on that PR (#8451) though, I think I'll change that to a json representation before updating this code.

Reply:
Basically the undocumented policy we've been using is `json` for machine-generated stuff, `yaml` for human-edited stuff (config files).  #2189 has a pretty lengthy discussion of how much faster `json` is.